## Renda mitja per barri, Asociacions i Serveis Socials 

In [13]:
import numpy as np
import json
import pandas as pd

import matplotlib.pyplot as plt
from scipy import stats
from sklearn import preprocessing

In [14]:
# Llegeixo el dataframe
#ServeisSocBCN = pd.read_csv('/Users/angelssastrenadal/Downloads/opendatabcn_llista-equipaments_serveis-socials-csv.csv', encoding='UTF-16')
ServeisSocBCN = pd.read_csv(r'C:\Users\34644\Desktop\JUMP2DIGITAL\Datasets\opendatabcn_llista-equipaments_serveis-socials-csv.csv', encoding='UTF-16')

pd.set_option('display.max_columns', None)
ServeisSocBCN

register_id                                               name  \
0     99400233227  Residència Assistida per a Gent Gran Amavir Horta   
1     92086034418                          Llar Residència Valldaura   
2     92086028301                        Club d'Atletisme Nou Barris   
3     99400620504                                      Centre de Dia   
4     99400620561                      Residència Llar Vila Olímpica   
...            ...                                                ...   
1223  99400714677                               Centre Obert Tronada   
1224  99400714678               Centre Obert Cel Bon Pastor Infantil   
1225  99400720536  Servei Especialitzat d'Atenció a la Infància i...   
1226  99400708884                       Centre Obert CEL Adolescents   
1227  99400718948                 Fundació Tutelar Taller Sant Jordi   

      institution_id                                   institution_name  \
0                NaN                                                NaN   
1                NaN                                                NaN   
2                NaN                                                NaN   
3       9.940027e+10  Residència Assistida Sanitas Residencial Conse...   
4                NaN                                                NaN   
...              ...                                                ...   
1223    1.269135e+09                         Associació Juvenil Tronada   
1224    9.940007e+10  Fundació Privada Centre Educatiu i de Lleure (...   
1225             NaN                                                NaN   
1226    9.940007e+10  Fundació Privada Centre Educatiu i de Lleure (...   
1227             NaN                                                NaN   

                               created                          modified  \
0            2009-10-06T13:14:33+02:00  2022-12-30T18:46:24.728176+01:00   
1            1991-03-07T00:00:00+01:00  2023-03-13T16:25:19.513894+01:00   
2            1988-10-18T00:00:00+01:00  2023-03-09T15:22:27.076168+01:00   
3            2019-02-18T00:00:00+01:00  2023-01-13T11:19:50.611665+01:00   
4            2019-02-19T14:30:08+01:00  2023-01-30T18:50:44.593844+01:00   
...                                ...                               ...   
1223  2023-06-09T11:36:57.644286+02:00  2023-06-09T11:42:13.411267+02:00   
1224  2023-06-09T11:44:44.230098+02:00  2023-06-09T11:48:39.247718+02:00   
1225  2023-09-20T14:06:16.390175+02:00  2023-09-20T14:32:56.935881+02:00   
1226  2023-03-27T11:24:04.325173+02:00  2023-03-27T11:27:37.122616+02:00   
1227  2023-08-16T12:27:26.381059+02:00  2023-09-27T10:13:27.519709+02:00   

      addresses_roadtype_id  addresses_roadtype_name  addresses_road_id  \
0                       NaN                      NaN           700812.0   
1                       NaN                      NaN           119003.0   
2                       NaN                      NaN           120807.0   
3                       NaN                      NaN            89004.0   
4                       NaN                      NaN            62609.0   
...                     ...                      ...                ...   
1223                    NaN                      NaN            39902.0   
1224                    NaN                      NaN           110508.0   
1225                    NaN                      NaN           100209.0   
1226                    NaN                      NaN            38309.0   
1227                    NaN                      NaN           700779.0   

          addresses_road_name  addresses_start_street_number  \
0               Av Can Marcet                           11.0   
1             Pg Fabra i Puig                          443.0   
2                Via Favència                          450.0   
3           C Consell de Cent                          210.0   
4                   Av Icària                          160.0   
...                       ...  

In [5]:
ServeisSocBCN.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1228 entries, 0 to 1227
Data columns (total 39 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   register_id                    1228 non-null   object 
 1   name                           1228 non-null   object 
 2   institution_id                 438 non-null    float64
 3   institution_name               438 non-null    object 
 4   created                        1228 non-null   object 
 5   modified                       1228 non-null   object 
 6   addresses_roadtype_id          0 non-null      float64
 7   addresses_roadtype_name        0 non-null      float64
 8   addresses_road_id              1204 non-null   float64
 9   addresses_road_name            1187 non-null   object 
 10  addresses_start_street_number  1187 non-null   float64
 11  addresses_end_street_number    58 non-null     float64
 12  addresses_neighborhood_id      1204 non-null   f

- Veig que nomes hi ha 1187 amb nom de districte i nom de barri, de moment escollim nomes aquests i 4 columnes

In [6]:
selected_columns = ['register_id', 'name','addresses_neighborhood_id','addresses_neighborhood_name', 'addresses_district_id', 'addresses_district_name']
ServeisSocBCNx = ServeisSocBCN.loc[:, selected_columns]

ServeisSocBCNx = ServeisSocBCNx.dropna()
ServeisSocBCNx

register_id                                               name  \
0     99400233227  Residència Assistida per a Gent Gran Amavir Horta   
1     92086034418                          Llar Residència Valldaura   
2     92086028301                        Club d'Atletisme Nou Barris   
3     99400620504                                      Centre de Dia   
4     99400620561                      Residència Llar Vila Olímpica   
...            ...                                                ...   
1223  99400714677                               Centre Obert Tronada   
1224  99400714678               Centre Obert Cel Bon Pastor Infantil   
1225  99400720536  Servei Especialitzat d'Atenció a la Infància i...   
1226  99400708884                       Centre Obert CEL Adolescents   
1227  99400718948                 Fundació Tutelar Taller Sant Jordi   

      addresses_neighborhood_id      addresses_neighborhood_name  \
0                          41.0                 la Vall d'Hebron   
1                          48.0                     la Guineueta   
2                          52.0                   la Prosperitat   
3                           8.0  l'Antiga Esquerra de l'Eixample   
4                          67.0    la Vila Olímpica del Poblenou   
...                         ...                              ...   
1223                       46.0              el Turó de la Peira   
1224                       59.0                    el Bon Pastor   
1225                       45.0                            Porta   
1226                       59.0                    el Bon Pastor   
1227                        1.0                         el Raval   

      addresses_district_id addresses_district_name  
0                       7.0          Horta-Guinardó  
1                       8.0              Nou Barris  
2                       8.0              Nou Barris  
3                       2.0                Eixample  
4                      10.0              Sant Martí  
...                     ...                     ...  
1223                    8.0              Nou Barris  
1224                    9.0             Sant Andreu  
1225                    8.0              Nou Barris  
1226                    9.0             Sant Andreu  
1227                    1.0            Ciutat Vella  

[1187 rows x 6 columns]

In [7]:
ServeisSocBCNx.describe()

addresses_neighborhood_id  addresses_district_id
count                1187.000000            1187.000000
mean                   29.631845               5.274642
std                    21.487792               2.914531
min                     1.000000               1.000000
25%                    10.000000               2.000000
50%                    27.000000               5.000000
75%                    43.000000               7.000000
max                    73.000000              10.000000

In [8]:
# Comprovem els tipus de variable per columna
column_data_types = ServeisSocBCNx.dtypes

# Iterem per columna i mirem el tipus de data
for column, dtype in column_data_types.items():
    print(f"Column '{column}' has dtype: {dtype}")

Column 'register_id' has dtype: object
Column 'name' has dtype: object
Column 'addresses_neighborhood_id' has dtype: float64
Column 'addresses_neighborhood_name' has dtype: object
Column 'addresses_district_id' has dtype: float64
Column 'addresses_district_name' has dtype: object


In [10]:
def create_json(json_path, dataset, type_function, grup_by = 'Codi_Bari', column_name = 'Valor'):
    import json
    # Assuming Escoles_infantils is your DataFrame
    grouped = dataset.groupby(grup_by)

    # Using dictionary comprehension
    new_data = {
        neighborhood: type_function(group[column_name].unique())
        for neighborhood, group in grouped
    }

    with open(json_path, 'w') as fp:
        json.dump(new_data, fp)




In [ ]:
create_json('serveissocbcnx.json', ServeisSocBCNx, list, 'addresses_neighborhood_id', 'name')

In [9]:
AssociacionsBCN = pd.read_csv('/Users/angelssastrenadal/Downloads/opendatabcn_llista-equipaments_associacions-csv.csv', encoding='UTF-16')
ServeisSocBCNx
pd.set_option('display.max_columns', None)
AssociacionsBCN

register_id                                               name  \
0     99400156664                                    Servei Solidari   
1     75990547699                 Asociación de Bares y Restaurantes   
2     75990547767                                  Lanonima Imperial   
3     99400346056                                   Cos de Bastoners   
4     99400217760  Associació Catalana de Professionals de l'Estr...   
...            ...                                                ...   
9577  99400720255                                   Fundación Hestia   
9578  99400719684                         Plataforma Veïnal SI_RAVAL   
9579  99400720206              Associació Juvenil Vaya Tela Trinitat   
9580  99400720614                           Associació CTNNB1 España   
9581  99400720436  Associació taSHORTfest - Festival Internaciona...   

      institution_id            institution_name  \
0                NaN                         NaN   
1                NaN                         NaN   
2                NaN                         NaN   
3       9.208602e+10  Esbart Català de Dansaires   
4                NaN                         NaN   
...              ...                         ...   
9577             NaN                         NaN   
9578             NaN                         NaN   
9579             NaN                         NaN   
9580             NaN                         NaN   
9581             NaN                         NaN   

                               created                          modified  \
0            2006-04-04T09:36:41+02:00  2023-02-13T18:34:45.087923+01:00   
1            1998-11-21T00:00:00+01:00  2023-10-03T10:37:46.566558+02:00   
2            1998-11-21T00:00:00+01:00  2023-10-03T10:37:46.483293+02:00   
3            2014-06-29T00:00:00+02:00  2021-03-03T22:14:42.476656+01:00   
4            2008-11-25T08:28:20+01:00  2022-09-17T07:07:53.825640+02:00   
...                                ...                               ...   
9577  2023-09-15T14:41:24.762018+02:00  2023-09-18T09:18:39.314891+02:00   
9578  2023-09-05T14:45:36.288302+02:00  2023-10-06T09:14:07.947547+02:00   
9579  2023-09-14T14:30:47.276005+02:00  2023-09-15T12:55:50.445452+02:00   
9580  2023-09-22T09:01:24.253730+02:00  2023-09-22T13:16:25.554887+02:00   
9581  2023-09-19T14:00:20.143916+02:00  2023-09-26T13:42:17.979601+02:00   

      addresses_roadtype_id  addresses_roadtype_name  addresses_road_id  \
0                       NaN                      NaN           296600.0   
1                       NaN                      NaN            89707.0   
2                       NaN                      NaN           221305.0   
3                       NaN                      NaN            36802.0   
4                       NaN                      NaN           177101.0   
...                     ...                      ...                ...   
9577                    NaN                      NaN           700013.0   
9578                    NaN                      NaN           282106.0   
9579                    NaN                      NaN           358707.0   
9580                    NaN                      NaN           209900.0   
9581                    NaN                      NaN            22400.0   

          addresses_road_name  addresses_start_street_number  \
0          C Sant Antoni Abat                           61.0   
1               C Constitució                           99.0   
2                     Morales                           21.0   
3                    C Beates                            2.0   
4                Via Laietana                           48.0   
...                       ...                            ...   
9577       Plaça de Karl Marx                            1.0   
9578  Carrer de la Riera Alta                           59.0   
9579  Carrer de Vicenç Montal                            2.0   
9580       Avinguda Meridiana                      

In [10]:
AssociacionsBCN.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9582 entries, 0 to 9581
Data columns (total 39 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   register_id                    9582 non-null   object 
 1   name                           9582 non-null   object 
 2   institution_id                 1221 non-null   float64
 3   institution_name               1221 non-null   object 
 4   created                        9582 non-null   object 
 5   modified                       9582 non-null   object 
 6   addresses_roadtype_id          0 non-null      float64
 7   addresses_roadtype_name        0 non-null      float64
 8   addresses_road_id              9374 non-null   float64
 9   addresses_road_name            9402 non-null   object 
 10  addresses_start_street_number  9390 non-null   float64
 11  addresses_end_street_number    1106 non-null   float64
 12  addresses_neighborhood_id      9414 non-null   f

Veig que de les 9582 entrades, nomes hi ha 9414 amb nom de districte i nom de barri, de moment escollim nomes aquests i 6 columnes

In [11]:
selected_columns = ['register_id', 'name','addresses_neighborhood_id','addresses_neighborhood_name', 'addresses_district_id', 'addresses_district_name']
AssociacionsBCNx = AssociacionsBCN.loc[:, selected_columns]

AssociacionsBCNx = AssociacionsBCNx.dropna()
AssociacionsBCNx

register_id                                               name  \
0     99400156664                                    Servei Solidari   
1     75990547699                 Asociación de Bares y Restaurantes   
2     75990547767                                  Lanonima Imperial   
3     99400346056                                   Cos de Bastoners   
4     99400217760  Associació Catalana de Professionals de l'Estr...   
...            ...                                                ...   
9577  99400720255                                   Fundación Hestia   
9578  99400719684                         Plataforma Veïnal SI_RAVAL   
9579  99400720206              Associació Juvenil Vaya Tela Trinitat   
9580  99400720614                           Associació CTNNB1 España   
9581  99400720436  Associació taSHORTfest - Festival Internaciona...   

      addresses_neighborhood_id            addresses_neighborhood_name  \
0                           1.0                               el Raval   
1                          16.0                             la Bordeta   
2                          19.0                              les Corts   
3                           4.0  Sant Pere, Santa Caterina i la Ribera   
4                           4.0  Sant Pere, Santa Caterina i la Ribera   
...                         ...                                    ...   
9577                       43.0                                  Horta   
9578                        1.0                               el Raval   
9579                       57.0                      la Trinitat Vella   
9580                       60.0                            Sant Andreu   
9581                       38.0                           la Teixonera   

      addresses_district_id addresses_district_name  
0                       1.0            Ciutat Vella  
1                       3.0          Sants-Montjuïc  
2                       4.0               Les Corts  
3                       1.0            Ciutat Vella  
4                       1.0            Ciutat Vella  
...                     ...                     ...  
9577                    7.0          Horta-Guinardó  
9578                    1.0            Ciutat Vella  
9579                    9.0             Sant Andreu  
9580                    9.0             Sant Andreu  
9581                    7.0          Horta-Guinardó  

[9373 rows x 6 columns]

In [ ]:
create_json('associacionsbcnx.json', AssociacionsBCNx, list, 'addresses_neighborhood_id', 'name')

In [14]:

RentaBCN = pd.read_csv('/Users/angelssastrenadal/Downloads/2021_atles_renda_bruta_persona.csv', encoding='UTF-8')
RentaBCN

Any  Codi_Districte Nom_Districte  Codi_Barri  \
0     2021               1  Ciutat Vella           1   
1     2021               1  Ciutat Vella           1   
2     2021               1  Ciutat Vella           1   
3     2021               1  Ciutat Vella           1   
4     2021               1  Ciutat Vella           1   
...    ...             ...           ...         ...   
1063  2021              10    Sant Martí          73   
1064  2021              10    Sant Martí          65   
1065  2021              10    Sant Martí          69   
1066  2021              10    Sant Martí          69   
1067  2021              10    Sant Martí          71   

                                         Nom_Barri  Seccio_Censal  \
0                                         el Raval              1   
1                                         el Raval              2   
2                                         el Raval              3   
3                                         el Raval              4   
4                                         el Raval              5   
...                                            ...            ...   
1063                           la Verneda i la Pau            143   
1064                                       el Clot            234   
1065  Diagonal Mar i el Front Marítim del Poblenou            235   
1066  Diagonal Mar i el Front Marítim del Poblenou            236   
1067                       Provençals del Poblenou            237   

      Import_Renda_Bruta_€  
0                    13657  
1                    10724  
2                    11017  
3                    13156  
4                    11426  
...                    ...  
1063                 15051  
1064                 19421  
1065                 17305  
1066                 17814  
1067                 15888  

[1068 rows x 7 columns]

In [15]:
RentaBCN.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1068 entries, 0 to 1067
Data columns (total 7 columns):
 #   Column                Non-Null Count  Dtype 
---  ------                --------------  ----- 
 0   Any                   1068 non-null   int64 
 1   Codi_Districte        1068 non-null   int64 
 2   Nom_Districte         1068 non-null   object
 3   Codi_Barri            1068 non-null   int64 
 4   Nom_Barri             1068 non-null   object
 5   Seccio_Censal         1068 non-null   int64 
 6   Import_Renda_Bruta_€  1068 non-null   int64 
dtypes: int64(5), object(2)
memory usage: 58.5+ KB


In [16]:
media_renta_por_barrio = RentaBCN.groupby('Codi_Barri')['Import_Renda_Bruta_€'].mean().reset_index()

media_renta_por_barrio


Codi_Barri  Import_Renda_Bruta_€
0            1          12133.380952
1            2          16996.555556
2            3          15986.545455
3            4          18077.538462
4            5          23084.850000
..         ...                   ...
68          69          25008.285714
69          70          13337.230769
70          71          19156.000000
71          72          18040.666667
72          73          16520.095238

[73 rows x 2 columns]

In [17]:
nuevo_df = RentaBCN[['Codi_Districte', 'Nom_Districte', 'Codi_Barri', 'Nom_Barri']].drop_duplicates()

# Combina los DataFrames en uno nuevo usando 'merge' para tener la media de renta por código de barrio
resultat_renta = nuevo_df.merge(media_renta_por_barrio, on='Codi_Barri', how='inner')
pd.set_option('display.max_rows', None)
resultat_renta

Codi_Districte        Nom_Districte  Codi_Barri  \
0                1         Ciutat Vella           1   
1                1         Ciutat Vella           2   
2                1         Ciutat Vella           3   
3                1         Ciutat Vella           4   
4                2           L'Eixample           5   
5                2           L'Eixample           6   
6                2           L'Eixample           7   
7                2           L'Eixample           8   
8                2           L'Eixample           9   
9                2           L'Eixample          10   
10               3       Sants-Montjuïc          11   
11               3       Sants-Montjuïc          12   
12               3       Sants-Montjuïc          13   
13               3       Sants-Montjuïc          14   
14               3       Sants-Montjuïc          15   
15               3       Sants-Montjuïc          16   
16               3       Sants-Montjuïc          17   
17               3       Sants-Montjuïc          18   
18               4            Les Corts          19   
19               4            Les Corts          20   
20               4            Les Corts          21   
21               5  Sarrià-Sant Gervasi          22   
22               5  Sarrià-Sant Gervasi          23   
23               5  Sarrià-Sant Gervasi          24   
24               5  Sarrià-Sant Gervasi          25   
25               5  Sarrià-Sant Gervasi          26   
26               5  Sarrià-Sant Gervasi          27   
27               6               Gràcia          28   
28               6               Gràcia          29   
29               6               Gràcia          30   
30               6               Gràcia          31   
31               6               Gràcia          32   
32               7       Horta-Guinardó          33   
33               7       Horta-Guinardó          34   
34               7       Horta-Guinardó          35   
35               7       Horta-Guinardó          36   
36               7       Horta-Guinardó          37   
37               7       Horta-Guinardó          38   
38               7       Horta-Guinardó          39   
39               7       Horta-Guinardó          40   
40               7       Horta-Guinardó          41   
41               7       Horta-Guinardó          42   
42               7       Horta-Guinardó          43   
43               8           Nou Barris          44   
44               8           Nou Barris          45   
45               8           Nou Barris          46   
46               8           Nou Barris          47   
47               8           Nou Barris          48   
48               8           Nou Barris          49   
49               8           Nou Barris          50   
50               8           Nou Barris          51   
51               8           Nou Barris          52   
52               8           Nou Barris          53   
53               8           Nou Barris          54   
54               8           Nou Barris          55   
55               8           Nou Barris          56   
56               9          Sant Andreu          57   
57               9          Sant Andreu          58   
58               9          Sant Andreu          59   
59               9          Sant Andreu          60   
60               9          Sant Andreu          61   
61               9          Sant Andreu          62   
62               9          Sant Andreu          63   
63              10           Sant Martí          64   
64              10           Sant Martí          65   
65              10           Sant Martí          66   
66              10           Sant Martí          67   
67              10           Sant Martí          68   
68              10           Sant Martí          69   
69              10           Sant Martí          70   
70              10           Sant Martí          71   
71              10           Sant Martí     

In [18]:

json_data = resultat_renta[['Codi_Barri', 'Import_Renda_Bruta_€']]


data = {str(k): v for k, v in zip(json_data['Codi_Barri'].values, json_data['Import_Renda_Bruta_€'].values)}


In [19]:
# Mostrar el JSON resultante
json.dump(data, open('renda.json', 'w'), indent=4)